homeassistant/components/assist_pipeline/pipeline.py

```diff
@@ -1565,21 +1572,19 @@ class PipelineRuns:
 
     def __init__(self, pipeline_store: PipelineStorageCollection) -> None:
         """Initialize."""
-        self._pipeline_runs: dict[str, list[PipelineRun]] = {}
+        self._pipeline_runs: dict[str, dict[str, PipelineRun]] = defaultdict(dict)
         self._pipeline_store = pipeline_store
         pipeline_store.async_add_listener(self._change_listener)
 
     def add_run(self, pipeline_run: PipelineRun) -> None:
         """Add pipeline run."""
         pipeline_id = pipeline_run.pipeline.id
-        if pipeline_id not in self._pipeline_runs:
-            self._pipeline_runs[pipeline_id] = []
-        self._pipeline_runs[pipeline_id].append(pipeline_run)
+        self._pipeline_runs[pipeline_id][pipeline_run.id] = pipeline_run
 
     def remove_run(self, pipeline_run: PipelineRun) -> None:
         """Remove pipeline run."""
         pipeline_id = pipeline_run.pipeline.id
-        self._pipeline_runs[pipeline_id].remove(pipeline_run)
+        self._pipeline_runs[pipeline_id].pop(pipeline_run.id, None)
```

Comment:
We shouldn't call this twice, so let it blow up if we accidentally do. Like before.
```suggestion
        self._pipeline_runs[pipeline_id].pop(pipeline_run.id)
```